src/forcing/ForcingsEngineLumpedDataProvider.cpp

```diff
@@ -0,0 +1,250 @@
+        static_cast<const int*>(bmi_->GetValuePtr("CAT-ID")),
+        id_dim
+    );
+
+    var_divides_.reserve(id_dim);
+    for (int i = 0; i < id_dim; ++i) {
+        var_divides_[cat_id[i]] = i;
+    }
+
+    var_cache_ = decltype(var_cache_){{ 2, id_dim, var_dim }};
+
+    // Cache initial iteration
+    update_value_storage_();
+    // this->increment_time(); // TODO: why???
+}
+
+BaseProvider* Provider::lumped_instance(
+    const std::string& init,
+    const std::string& time_begin,
+    const std::string& time_end,
+    const std::string& time_fmt
+)
+{
+    auto time_begin_epoch = static_cast<size_t>(parse_time(time_begin, time_fmt));
+    auto time_end_epoch = static_cast<size_t>(parse_time(time_end, time_fmt));
+
+
+    auto provider = std::unique_ptr<ForcingsEngineLumpedDataProvider>{
+        new Provider{init, time_begin_epoch, time_end_epoch}
+    };
+
+    return set_instance(init, std::move(provider));
+}
+
+std::size_t Provider::divide_index(const std::string& divide_id) noexcept
+{
+    const auto  id_sep   = divide_id.find('-');
+    const char* id_split = id_sep == std::string::npos
+                           ? &divide_id[0]
+                           : &divide_id[id_sep + 1];
+    const int   id_int   = std::atoi(id_split);
+
+    auto pos = var_divides_.find(id_int);
+    if (pos == var_divides_.end()) {
+        return bad_index;
+    }
+
+    // implicit cast to unsigned
+    return pos->second;
+}
+
+std::size_t Provider::variable_index(const std::string& variable) noexcept
+{
+    auto vars = this->get_available_variable_names();
+    const auto* pos = std::find_if(
+        vars.begin(),
+        vars.end(),
+        [&](const std::string& var) -> bool {
+            // Checks for a variable name, plus the case where only the prefix is given, i.e.
+            // both variable_index("U2D") and variable_index("U2D_ELEMENT") will work.
+            return var == variable || var == (variable + "_ELEMENT");
+        }
+    );
+    
+    if (pos == vars.end()) {
+        return bad_index;
+    }
+
+    // implicit cast to unsigned
+    return std::distance(vars.begin(), pos);
+}
+
+bool Provider::next()
+{
+    bmi_->Update();
+    this->increment_time();
+    this->update_value_storage_();
+    return true;
+}
+
+void Provider::update_value_storage_()
+{
+    const auto outputs = this->get_available_variable_names();
+    const auto size = var_divides_.size();
+    for (std::size_t vi = 0; vi < outputs.size(); vi++) {
+        // Get current values for each variable
+        const auto var_span = boost::span<const double>{
+            static_cast<double*>(bmi_->GetValuePtr(outputs[vi])),
+            size
+        };
+
+        for (std::size_t di = 0; di < size; di++) {
+            // Move 0 time (current) to 1 (previous), and set the
+            // current values from var_span
+            var_cache_.at({{1, di, vi}}) = var_cache_.at({{0, di, vi}});
+            var_cache_.at({{0, di, vi}}) = var_span[di];
```

Comment:
Not a request to change anything now, but for your awareness, there's a common pattern in HPC codes for old/new values of an array to avoid copying. It's to have two storage instances, and swap (or literally `swap`) between them. If we end up wanting/needing to optimize this, that's probably what we'd implement.